src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -264,17 +264,38 @@ bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock
     assert(domBlock->bbJumpKind == BBJ_COND);
 
     // If the dominating block is not the immediate dominator
-    // we would need to duplicate a lot of code to thread
-    // the jumps. Pass for now.
+    // we might need to duplicate a lot of code to thread
+    // the jumps. See if that's the case.
     //
-    if (domBlock != block->bbIDom)
+    const bool isIDom = domBlock == block->bbIDom;
+    if (!isIDom)
     {
-        JITDUMP(" -- not idom, so no threading\n");
-        return false;
+        // Walk up the dom tree until we hit dom block.
+        //
+        // If none of the doms in the stretch are BBJ_COND,
+        // then we must have already optimized them, and
+        // so should not have to duplicate code to thread.
+        //
+        BasicBlock* idomBlock = block->bbIDom;
+        while ((idomBlock != nullptr) && (idomBlock != domBlock))
+        {
+            if (idomBlock->bbJumpKind == BBJ_COND)
+            {
+                JITDUMP(" -- " FMT_BB " not closest branching dom, so no threading\n", idomBlock->bbNum);
+                return false;
+            }
+            JITDUMP(" -- bypassing %sdom " FMT_BB " as it was already optimized\n",
+                    (idomBlock == block->bbIDom) ? "i" : "", idomBlock->bbNum);
+            idomBlock = idomBlock->bbIDom;
+        }
+
+        // If we didn't bail out above, we should have reached domBlock.
+        //
+        assert(idomBlock == domBlock);
```

Comment:
Do we need to check that the new dominator is in the same EH region here?

Reply:
Yes. 

EH region constraints are checked later on, when `optJumpThread` examines exactly how control leaving `domBlock` reaches the block to be optimized.